### PR TITLE
[P2P] Fix flaky P2P tests

### PR DESCRIPTION
**Pull request by bryanchriswhite**
## Description

There were two tests which were expressing flaky behavior. In both cases the tests failed via panic, which is my hypothetical explanation for why nothing printed in the test failure output; I think that CI step is not sensitive to the panic failure mode output.

To reproduce and/or verify this change, use the `-count=X` and `-run` flags to run a specific test `X` number of times consecutively:

```bash
go test -count=10 -run Libp2pKademliaPeerDiscovery ./p2p/background

go test -count=10 -run BackgroundRouter_broadcast ./p2p/background
```

### `TestLibp2pKademliaPeerDiscovery`

Wasn't calling `#Close()` on its libp2p `Host` or it's DHT. This has been added in a `t.Cleanup(...)`.

### `BackgroundRouter_broadcast`

Was panicking when an unanticipated connection is made in the context of kademlia overlay network bootstrapping (which is effectively out of our control). The test now recovers from the panic, ignores this specific error, and fails the test otherwise (any other error).

## Issue

N/A

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [x] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- handle flaky negative WaitGroup panic in background router broadcast test
- handle flaky kad baseline test by adding proper cleanup

## Testing

- [x] `make develop_test`; if any code changes were made
- [ ] `make test_e2e` on [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any code changes were made
- [ ] `e2e-devnet-test` passes tests on [DevNet](https://pocketnetwork.notion.site/How-to-DevNet-ff1598f27efe44c09f34e2aa0051f0dd); if any code was changed
- [ ] [Docker Compose LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md); if any major functionality was changed or introduced
- [ ] [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any infrastructure or configuration changes were made

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [x] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: p2p/background/kad_discovery_baseline_test.go [hunk 1]
```diff
@@ -50,9 +50,21 @@ func TestLibp2pKademliaPeerDiscovery(t *testing.T) {
 	require.ElementsMatchf(t, expectedPeerIDs, host3.Peerstore().Peers(), "host3 peer IDs don't match")
 
 	// add another peer to network...
-	addr4, host4, _ := setupHostAndDiscovery(t, ctx, defaults.DefaultP2PPort+3, bootstrapAddr)
+	addr4, host4, kad4 := setupHostAndDiscovery(t, ctx, defaults.DefaultP2PPort+3, bootstrapAddr)
 	expectedPeerIDs = append(expectedPeerIDs, host4.ID())
 
+	t.Cleanup(func() {
```
Comment (h5law) [inline p2p/background/kad_discovery_baseline_test.go:56]: Nice catch here 👍🏼

**Review by h5law — APPROVED** [review]
Changes look good to me, I think the reconnection issue will take some time to figure out how best to control this behaviour on our end. But lets :shipit:

File: p2p/background/router_test.go [hunk 2]
```diff
@@ -303,6 +303,23 @@ func TestBackgroundRouter_Broadcast(t *testing.T) {
 	// setup notifee/notify BEFORE bootstrapping
 	notifee := &libp2pNetwork.NotifyBundle{
 		ConnectedF: func(_ libp2pNetwork.Network, _ libp2pNetwork.Conn) {
+			// TECHDEBT: it's rare but possible that a host will re-connect,
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply p2p/background/router_test.go:306]: Agreed. There's no reason we have to design the test this way and for this reason it's turned out to be an incompatible design given our quality tolerance. :sweat_smile:

File: p2p/background/kad_discovery_baseline_test.go [hunk 3]
```diff
@@ -50,9 +50,21 @@ func TestLibp2pKademliaPeerDiscovery(t *testing.T) {
...
+		for _, kad := range []*dht.IpfsDHT{kad1, kad2, kad3, kad4} {
```
Comment (Olshansk) [inline p2p/background/kad_discovery_baseline_test.go:57]: 🆒 inline array

File: p2p/background/router_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply p2p/background/router_test.go:306]: Appreciate the detailed TECHDEBT comment 👍

**Review by Olshansk — APPROVED** [review]

**Pull request merged**